### DISCUSSION TITLE
Stored energy kernel with mesh adaptivity

### DISCUSSION THREAD

**Discussion by HiltyF** [idaholab/moose#24600] [Q&A General]
Hello all,

I am working on a recrystallization problem and noticed that the `ACSEDGPoly` kernel is heavily influenced by mesh adaptivity. The effect is seen in the GIF below, both cases are  a single circular grain in the middle of a large grain with high dislocation density. The Small grain grows due to the dislocation energy, as expected. However, when mesh adaptivity is on (top right) the circle grain becomes rapidly deformed and grows significantly slower compared to when the mesh stays uniform (bottom right). The contours just highlight the differences in the circular grains.

The input file for these cases is at the end of the post. If I comment out the `ACSEDGPoly` kernels and perform this test again then the results with and without mesh adaptivity are the same. 

Can anyone give insight into what is causing this behavior?

Thanks,
Floyd

![Stored_Energy_Adaptivity](https://github.com/idaholab/moose/assets/122134959/5e723135-733b-4a63-adb1-9675727418d6)

```
file = 'Stored_Energy_Adaptivity'

[Mesh]
  parallel_type = 'REPLICATED'
  type = GeneratedMesh
  dim = 2
  nx = 50
  ny = 50
  xmin = -5e-4
  xmax = 5e-4
  ymin = -5e-4
  ymax = 5e-4
  elem_type = QUAD4
[]

[Adaptivity]
  steps = 1
  marker = gr1_mark
  max_h_level = 1
  initial_steps = 1
  initial_marker = uniform
  [Markers]
    [gr1_mark]
      type = ValueRangeMarker
      variable = gr1
      lower_bound = 0.01
      upper_bound = .95
    []
    [uniform]
      type = UniformMarker
      mark = REFINE
    []
  []
[]

[GlobalParams]
  op_num = 2
  var_name_base = gr
  enable_jit = false
[]

[UserObjects]
  [grain_tracker]
    type = GrainTracker
    compute_var_to_feature_map = true
    threshold = 0.2
    connecting_threshold = 0.08
    compute_halo_maps = true
    remap_grains = true
  []
[]

[ICs]
  [IC_gr0]
    type = SpecifiedSmoothCircleIC
    variable = gr0
    x_positions = 0
    y_positions = 0
    z_positions = 0
    radii = 1e-4
    int_width = 1e-5
    invalue = 0
    outvalue = 1
  []
  [IC_gr1]
    type = SpecifiedSmoothCircleIC
    variable = gr1
    x_positions = 0
    y_positions = 0
    z_positions = 0
    radii = 1e-4
    int_width = 1e-5
    invalue = 1
    outvalue = 0
  []
[]

[Variables]
  [gr0]
    order = FIRST
    family = LAGRANGE
  []
  [gr1]
    order = FIRST 
    family = LAGRANGE
  []
[]

[AuxVariables]
  [bnds]
    order = FIRST
    family = LAGRANGE
  []
  [unique_grains]
    order = CONSTANT
    family = MONOMIAL
  []
  [T]
    [InitialCondition]
      type = ConstantIC
      value = 1200
    []
  []
[]

[Kernels]
  [PolycrystalKernel]
  []
  # [PolycrystalStoredEnergy]
  #   grain_tracker = grain_tracker
  #   deformed_grain_num = 1
  # []
  [ACStoredEnergy_gr0]
    type = ACSEDGPoly
    deformed_grain_num = 1
    grain_tracker = grain_tracker
    op_index = 0
    v = gr1
    variable = gr0
  []
  [ACStoredEnergy_gr1]
    type = ACSEDGPoly
    deformed_grain_num = 1
    grain_tracker = grain_tracker
    op_index = 1
    v = gr0
    variable = gr1
  []
[]

[AuxKernels]
  [unique_grains]
    type = FeatureFloodCountAux
    variable = unique_grains
    flood_counter = grain_tracker
    field_display = UNIQUE_REGION
    execute_on = 'INITIAL TIMESTEP_END'
  []
  [bnds_aux]
    type = BndsCalcAux
    variable = bnds
    execute_on = timestep_end
  []
[]

[Materials]
  [consts]
    type = GenericConstantMaterial
    #              J/m3   m          -     -         J/K/mol  m          eV/K        eV/K
    prop_names = ' shear  b          alpha gamma Q_a R        gb_length  gamma_asymm kb          J2eV          length time'
    prop_values = '7.6e10 1.7983e-10 0.5   1     275 8.3145   5e-8       1.5         8.617343e-5 6.24150974e18 1e-3   1'
  []
  [gb_scale]
    type = ParsedMaterial
    property_name = lgb
    expression = 'gb_length/length'
    material_property_names = 'gb_length length'
  []
  [dislocation_density]
    type = ParsedMaterial
    ## 1/m^2
    property_name = Disloc_Den_i
    expression = 'gr0*7e15*length^2'
    coupled_variables = 'gr0'
    material_property_names = 'length'
    outputs = 'exodus'
  []

  ## Grain Growth
  [GBenergy]
    type = ParsedMaterial
    ## J/m^2 -> eV/m^2
    property_name = 'GBenergy'
    expression = '(1/(2*a^2*N)*(R*T*log(D0b/D0l)+(Hl-Hb)))*J2eV*length^2' # Varin  https://doi.org/10.1016/0025-5416(84)90144-7
    #                       m         1/mol    m^2/s  m^2/s     J/mol J/mol
    constant_names = '      a         N        D0b    D0l       Hl    Hb    '
    constant_expressions = '3.598e-10 6.022e23 6.1e-4 0.0118e-4 280   177.52'
    coupled_variables = 'T'
    material_property_names = 'R J2eV length'
  []
  [kappa_op]
    type = ParsedMaterial
    ## eV/mm
    property_name = 'kappa_op'
    expression = '3/4*GBenergy*lgb' # Aagesen 2018
    material_property_names = 'GBenergy lgb'
    # outputs = 'exodus'
  []
  [kappa]
    type = ParsedMaterial
    ## eV/mm
    property_name = 'kappa'
    expression = '3/4*GBenergy*lgb' # Aagesen 2018
    material_property_names = 'GBenergy lgb'
    # outputs = 'exodus'
  []
  [mu]
    type = ParsedMaterial
    ## eV/mm^3
    property_name = 'mu'
    expression = '6*GBenergy/lgb' # Aagesen 2018
    material_property_names = 'GBenergy lgb'
    # outputs = 'exodus'
  []
  [GBmob]
    type = ParsedMaterial
    ## Units m/GPa/s -> m/Pa/s (Pa = kg/m/s^2) -> m^2*s/kg (m^4/J/s)
    property_name = 'GBmob'
    expression = '(0.0000000669538*T^3-0.000110855*T^2+0.0853373*T-17.4132)/GPa2Pa'
    constant_names = 'GPa2Pa'
    constant_expressions = '1e9  '
    coupled_variables = 'T'
  []
  [Mgb]
    type = ParsedMaterial
    ## m^4/J/s -> m^4/eV/s
    property_name = Mgb
    expression = 'GBmob*time/J2eV/length^4'
    material_property_names = 'GBmob time J2eV length'
    # outputs = 'exodus'
  []
  [L]
    type = ParsedMaterial
    ## m^3/eV/s
    property_name = L
    expression = 'Mgb*GBenergy/kappa'
    material_property_names = 'Mgb GBenergy kappa lgb'
    # outputs = 'exodus'
  []
  [rho_eff]
    type = ParsedMaterial
    # 1/m^2
    property_name = rho_eff
    expression = 'Disloc_Den_i'
    material_property_names = 'Disloc_Den_i'
    # outputs = 'exodus'
  []
  [Stored_free_energy]
    type = DerivativeParsedMaterial
    ## J/m -> eV/m
    property_name = beta
    expression = '0.5*shear*b^2*J2eV*length'
    material_property_names = 'shear b J2eV length'
    outputs = 'exodus'
  []
[]

[Preconditioning]
  [ASM]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -ksp_grmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
    petsc_options_value = ' asm      31                  preonly       lu           2'
  []
[]

[Executioner]
  type = Transient
  solve_type = PJFNK
  l_max_its = 15
  l_tol = 1e-7
  nl_max_its = 20
  nl_abs_tol = 1e-11
  nl_rel_tol = 1e-8
  start_time = 0.0
  end_time = 1
  num_steps = 100
  [TimeStepper]
    type = IterationAdaptiveDT
    dt = 1e-7
    optimal_iterations = 4
    cutback_factor = 0.8
    growth_factor = 1.2
  []
[]

[Postprocessors]
  [0_dt]
    type = TimestepSize
  []
  [0_run_time]
    type = PerfGraphData
    section_name = "Root"
    data_type = total
  []
  [1_dofs]
    type = NumDOFs
  []
  [2_aveGrainSize]
    type = AverageGrainVolume
    feature_counter = grain_tracker
  []
[]

[Outputs]
  [out]
    file_base = CSV/${file}
    type = CSV
  []
  [exodus]
    type = Exodus
    interval = 1
    file_base = Exodus/${file}
    execute_on = 'INITIAL TIMESTEP_END FINAL'
  []
[]
```

**Comment by HiltyF** [comment]
Just poking this since it seems to have fallen off, I am really curious if anyone knows why mesh adaptivity would cause such a strong change in the results. I feel like I have the mesh refinement adequately surrounding the grain boundary such that the difference should be small, if at all.

**Reply by GiudGiud** [reply]
I would speculate that mesh adaptivity is not a conservative process. Refining the mesh is not conserving the quantities solved by your physics, creating discrepancies

But I m not an expert on this. Let's ping some real ones:
@permcody @laagesen

**Comment by laagesen** [comment]
I think the parameters you are using in the `ValueRangeMarker` seem reasonable, but I have not seen that marker used for phase-field modeling before. It seems like it is doing what one would expect, but I would try using the `GradientJumpIndicator` with `ErrorFractionMarker` that we have more typically used in phase-field modeling to see if the same thing happens.

**Comment by HiltyF** [comment]
Thank you for the suggestion, @laagesen. I was able to find a combination of the coarsen and refine values in the `ErrorFractionMarker` that maintains the correct shape for the growing grain, however its growth rate was significantly reduced. There seems to be a threshold for the coarsen value below which the behavior is consistent, however the refine fraction is more touchy. It seems that around refine = 0.7 I get the best behavior, lower will slow the growth more and there are some noticeable fluctuations, higher values cause faster growth but it is extremely erratic (as shown below). 

Do you have any suggestions about how to choose values for coarsen and refine? Also, do you have any other tips for how to improve the behavior, or what is the root cause for the high level of mesh adaptivity sensitivity?

Thanks again for your help!

Red: Uniform mesh
Blue: Refine = 0.4
Black: Refine = 0.7
Green: Refine = 0.99

![Marker_Indicator_Adaptivity](https://github.com/idaholab/moose/assets/122134959/cb83d3d7-b53b-4537-9470-1624978b2137)

**Comment by laagesen** [comment]
My gut feel is that the dislocation stored energy driving force must be really large compared to the curvature driving force, and you need a high mesh resolution to capture this properly. Did you ever try to check mesh size sensitivity on a uniform mesh prior to starting to add adaptivity? Just to make sure that the finest resolution you are using in the uniform mesh is actually fine enough to obtain a converged result. Maybe you need a finer mesh that would typically needed to resolve the interface width, which would normally be the limiting factor. This is all I can think of off hand...

**Comment by HiltyF** [comment]
Thank you again for the suggestions.

I did a mesh size sensitivity study on a uniform mesh, the mesh size I am using gives the same results as more refined meshes. I also tried increasing the number of refinement levels for the adaptive mesh, up to having ~20 elements across the interface. In these adaptive cases the growth changes with any change to the adaptivity scheme, but the growth behavior remains erratic.

With regard to the dislocation stored energy driving force, I lowered the dislocation density until the circle grain no longer grew. Just above that, when the stored energy driving force is just slightly larger than the curvature driving force, the grain still grows erratically. So, the culprit does not appear to be a large difference in driving forces.

One other thing, @GiudGiud, has suggested that it could be due to a lack of conservation. Since dislocations are not conserved, and that determines the stored energy driving force, it's difficult to really investigate. The best I could think of was lower the phase mobility to where the grains no longer evolve, in this situation as the mesh adapts the integral of the dislocation density does not change. So, I don't believe the adaptivity is a source for non-conservation, but I am open to additional ways I could try to verify this.
